jobs/payment-jobs/tasks/bcol_refund_confirmation_task.py

```diff
@@ -0,0 +1,110 @@
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Task to update refunded invoices that have been processed by BCOL."""
+from __future__ import annotations
+
+from decimal import Decimal
+from typing import Dict, List
+
+from flask import current_app
+from pay_api.models import Invoice, InvoiceReference, Payment, db
+from pay_api.models.payment_system import PaymentSystem as PaymentSystemModel
+from pay_api.utils.enums import InvoiceStatus, PaymentMethod, PaymentSystem
+from sentry_sdk import capture_message
+
+from services import oracle_db
+
+
+class BcolRefundConfirmationTask:  # pylint:disable=too-few-public-methods
+    """Task to update refunded invoices that have been processed by BCOL."""
+
+    @classmethod
+    def update_bcol_refund_invoices(cls):
+        """Update BCOL invoices that are in the REFUND_REQUESTED state to REFUNDED/CREDITED if they have been processed.
+
+        Steps:
+        1. get BCOL invoices in pay-db that are in REFUND_REQUESTED state
+        2. get BCOL records in colin-db (CPRD) for the invoices from step 1
+        3. compare the invoices in pay-db vs bcol records in colin-db
+          - update invoices to REFUNDED in pay-db that have the correct refund value in colin-db
+          - skip invoices that do not have a refund value in colin-db
+          - send a sentry message containing all refunds with mismatch values between pay-db and colin-db
+        """
+        invoice_refs = cls._get_paydb_invoice_refs_for_update()
+        if invoice_refs:
+            bcol_refund_records = cls._get_colin_bcol_records_for_invoices(invoice_refs)
+            current_app.logger.debug('BCOL refunded invoice numbers: %s', bcol_refund_records)
+
+            if bcol_refund_records:
+                cls._compare_and_update_records(invoice_refs, bcol_refund_records)
+        else:
+            current_app.logger.debug('No BCOL refunds to confirm.')
+
+    @classmethod
+    def _get_paydb_invoice_refs_for_update(cls) -> List[InvoiceReference]:
+        """Get outstanding refund requested BCOL invoice references."""
+        current_app.logger.debug('Collecting refund requested BCOL invoices...')
+        return db.session.query(InvoiceReference) \
+            .join(Invoice, Invoice.id == InvoiceReference.invoice_id) \
+            .join(Payment, Payment.invoice_number == InvoiceReference.invoice_number) \
+            .filter(Payment.payment_system.has(PaymentSystemModel.code == PaymentSystem.BCOL.value)) \
```

Comment:
Probably better to filter by `payment_method_code`, instead of Payment.payment_system - which would make it more consistent with the other jobs